typescript/phoenix-sdk/src/utils/market.ts

```diff
@@ -861,3 +862,64 @@ export function getMarketExpectedOutAmount({
 
   return expectedUnitsReceived;
 }
+
+/**
+ * Returns the expected amount in for a given desired amount out
+ *
+ * @param ladder
+ * @param takerFeeBps
+ * @param side If this is Bid, then the outAmount represents the number of base units desired; the returned result is the number of quote units required to get the desired amount of base units.
+ * If this is Ask, then the outAmount represents the number of quote units desired; the returned result is the number of base units requires to get the desired amount of quote units.
+ * @param outAmount
+ *
+ */
+export function getMarketExpectedInAmount({
+  ladder,
+  takerFeeBps,
+  side,
+  outAmount,
+}: {
+  ladder: UiLadder;
+  takerFeeBps: number;
+  side: Side;
+  outAmount: number;
+}): number {
+  let remainingUnitsOut = outAmount;
+  let expectedUnitsIn = 0;
+
+  if (side === Side.Bid) {
+    // The user wants outAmount of base units, so we need to calculate the number of quote units that will be required.
+    // We can do this by iterating through the asks and calculating the total base units available and at what price.
+    for (const [priceInQuoteUnitsPerBaseUnit, sizeInBaseUnits] of ladder.asks) {
+      if (sizeInBaseUnits > remainingUnitsOut) {
+        expectedUnitsIn += remainingUnitsOut * priceInQuoteUnitsPerBaseUnit;
+        remainingUnitsOut = 0;
+        break;
+      } else {
+        expectedUnitsIn += sizeInBaseUnits * priceInQuoteUnitsPerBaseUnit;
+        remainingUnitsOut -= sizeInBaseUnits;
+      }
+    }
+  } else {
+    // The user wants outAmount of quote units, so we need to calculate the number of base units that will be required.
+    // We can do this by iterating through the bids and calculating the total base units available and at what price.
+    for (const [priceInQuoteUnitsPerBaseUnit, sizeInBaseUnits] of ladder.bids) {
+      const totalQuoteUnitsAvailable =
+        sizeInBaseUnits * priceInQuoteUnitsPerBaseUnit;
+      if (totalQuoteUnitsAvailable > remainingUnitsOut) {
+        // Remaining quote units desired divided by price equals the necessary base units in.
+        expectedUnitsIn += remainingUnitsOut / priceInQuoteUnitsPerBaseUnit;
+        remainingUnitsOut = 0;
+        break;
+      } else {
+        expectedUnitsIn += sizeInBaseUnits;
+        remainingUnitsOut -= totalQuoteUnitsAvailable;
+      }
+    }
+  }
+
+  // Account for taker fees
+  expectedUnitsIn /= 1 - takerFeeBps / 10000;
```

Comment:
not sure if this is correct, or should be `*= (1 + takerFeeBps / 10000)`

it's probably different from buy/sell

although these numbers are *basically* the same if takerFeeBps is small

Reply:
Resolved offline.